internal/shim-sgx/src/handler/mod.rs

```diff
@@ -49,32 +52,217 @@ const OP_CPUID: u16 = 0xa20f;
+    }
+}
+
+impl guest::Handler for Handler<'_> {
+    fn block(&self) -> &[usize] {
+        self.block
+    }
+
+    fn block_mut(&mut self) -> &mut [usize] {
+        self.block
+    }
+
+    fn thread_local_storage(&mut self) -> &mut ThreadLocalStorage {
+        static mut TLS: ThreadLocalStorage = ThreadLocalStorage::new();
+        // FIXME: proper TLS implementation https://github.com/enarx/enarx/issues/1476
+        unsafe { &mut TLS }
+    }
+
+    fn arch_prctl(&mut self, code: c_int, addr: c_ulong) -> sallyport::Result<()> {
+        // TODO: Check that addr in %rdx does not point to an unmapped address
+        // and is not outside of the process address space.
+        match code {
+            ARCH_SET_FS => self.ssa.gpr.fsbase = addr,
+            ARCH_SET_GS => self.ssa.gpr.gsbase = addr,
+            ARCH_GET_FS => return Err(libc::ENOSYS),
+            ARCH_GET_GS => return Err(libc::ENOSYS),
+            _ => return Err(libc::EINVAL),
+        }
+        Ok(())
+    }
+
+    fn brk(&mut self, addr: Option<NonNull<c_void>>) -> sallyport::Result<NonNull<c_void>> {
+        Ok(NonNull::new(
+            crate::heap::HEAP
+                .write()
+                .brk(addr.map(|v| v.as_ptr() as usize).unwrap_or(0)) as _,
+        )
+        .unwrap())
+    }
+
+    fn madvise(
+        &mut self,
+        _addr: NonNull<c_void>,
+        _length: size_t,
+        _advice: c_int,
+    ) -> sallyport::Result<()> {
+        Ok(())
+    }
+
+    // Until EDMM, we can't change any page permissions.
+    // What you get is what you get. Fake success.
+    fn mprotect(
+        &mut self,
+        _addr: NonNull<c_void>,
+        _len: size_t,
+        _prot: c_int,
+    ) -> sallyport::Result<()> {
+        Ok(())
+    }
+
+    fn mmap(
+        &mut self,
+        addr: Option<NonNull<c_void>>,
+        length: size_t,
+        prot: c_int,
+        flags: c_int,
+        fd: c_int,
+        offset: off_t,
+    ) -> sallyport::Result<NonNull<c_void>> {
+        Ok(NonNull::new(crate::heap::HEAP.write().mmap::<libc::c_void>(
+            addr.map(|v| v.as_ptr() as usize).unwrap_or(0),
+            length,
+            prot,
+            flags,
+            fd,
+            offset,
+        )?)
+        .unwrap())
+    }
+
+    fn munmap(&mut self, addr: NonNull<c_void>, length: size_t) -> sallyport::Result<()> {
+        crate::heap::HEAP
+            .write()
+            .munmap::<libc::c_void>(addr.as_ptr() as _, length)
+    }
+}
+
+#[inline]
+unsafe fn as_aligned_ref<'a, T>(ptr: usize) -> Option<&'a T> {
+    let ptr = ptr as *const u8;
+    if ptr.align_offset(core::mem::align_of::<T>()) != 0 {
+        return None;
+    }
+    let ptr = ptr as *const T;
+
+    ptr.as_ref()
+}
+
+#[inline]
+unsafe fn as_aligned_mut<'a, T>(ptr: usize) -> Option<&'a mut T> {
+    let ptr = ptr as *const u8;
+    if ptr.align_offset(core::mem::align_of::<T>()) != 0 {
+        return None;
+    }
+    let ptr = ptr as *mut T;
+
+    ptr.as_mut()
+}
```

Comment:
These are duplicated and should be extracted in some shared utility library (or let's file a tech debt issue for it)